providers/implementations/signature/dsa_sig.c

```diff
@@ -438,6 +440,17 @@ static void *dsa_dupctx(void *vpdsactx)
             goto err;
     }
 
+    if (srcctx->aid != NULL
+        && srcctx->aid_len != 0
+        && srcctx->aid_len <= OSSL_MAX_ALGORITHM_ID_SIZE) {
+        size_t offset = srcctx->aid - srcctx->aid_buf;
+        if (offset < OSSL_MAX_ALGORITHM_ID_SIZE
+            && (offset + srcctx->aid_len) <= OSSL_MAX_ALGORITHM_ID_SIZE) {
+            dstctx->aid = dstctx->aid_buf + offset;
+            dstctx->aid_len = srcctx->aid_len;
+        }
+    }
+
```

Comment:
I'd suggest instead of doing this pointer arithmetics to `memcpy(dstctx->aid_buf, srcctx->aid, srcctx->aid_len);` and set `dstctx->aid = dstctx->aid_buf`;

Reply:
An alternative possibility is to call `memmove` on the `aid` value immediately after it is created to move it to the start of aid_buf. You can then drop the `aid` field altogether and no special handling needs to take place in the dup operation at all.

Reply:
Good, that's even better. I assume then `aid_len` would be used as the indicator whether an aid was written to the `aid_buf` or not. @lan1120 would you please consider doing this? Or, alternatively I can do that.

Reply:
OK, I'll do this.